### PR TITLE
[SPARK-31308][PySpark] Merging pyFiles to files argument for Non-PySpark applications

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This PR (SPARK-31308) proposed to add python dependencies even it is not Python applications.

### Why are the changes needed?

For now, we add `pyFiles` argument to `files` argument only for Python applications, in SparkSubmit. Like the reason in #21420, "for some Spark applications, though they're a java program, they require not only jar dependencies, but also python dependencies.", we need to add `pyFiles` to `files` even it is not Python applications.

### Does this PR introduce any user-facing change?

Yes. After this change, for non-PySpark applications, the Python files specified by `pyFiles` are also added to `files` like PySpark applications.

### How was this patch tested?

Manually test on jupyter notebook or do `spark-submit` with `--verbose`.

```
Spark config:
...
(spark.files,file:/Users/dongjoon/PRS/SPARK-PR-28077/a.py)
(spark.submit.deployMode,client)
(spark.master,local[*])
```

### CODE REVIEW

**Comment by viirya** [comment]
Maybe we need a unit test too. But let me wait for some comments first.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -474,10 +474,12 @@ private[spark] class SparkSubmit extends Logging {
         args.mainClass = "org.apache.spark.deploy.PythonRunner"
         args.childArgs = ArrayBuffer(localPrimaryResource, localPyFiles) ++ args.childArgs
       }
-      if (clusterManager != YARN) {
-        // The YARN backend handles python files differently, so don't merge the lists.
-        args.files = mergeFileLists(args.files, args.pyFiles)
-      }
+    }
+
+    // Non-PySpark applications will also need Python dependencies.
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: nit: `will also` -> `can`

**Comment by HyukjinKwon** [comment]
I think it's fine. cc @vanzin and @jerryshao

**Comment by dongjoon-hyun** [comment]
Hi, @viirya . This PR title looks too broad. Could you be more specific by excluding the scope of SPARK-24377 ?
> [SPARK-24377][Spark Submit] make --py-files work in non pyspark application

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -474,10 +474,12 @@ private[spark] class SparkSubmit extends Logging {
...
+    if (deployMode == CLIENT && clusterManager != YARN) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Just a question. Did we `mergeFileLists` for `deployMode != CLIENT` already?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 1 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: fixed.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 2 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: No, I don't see we merge it for cluster deploy mode.

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Got it. Thanks, @viirya .

**Comment by viirya** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @viirya and @HyukjinKwon .

**Pull request closed without merging**